File: MasteringDart/lib/chapters/chapter17/intermediate/intermediate_design_patterns.dart
// ignore_for_file: avoid_print
//
//  intermediate_design_patterns.dart
//  MasteringDart
//
//  Capítulo 17 — Patrones de diseño.
//  Nivel INTERMEDIO: Builder — construir User paso a paso con API fluida.
//
//  Ejecutar con: dart run lib/chapters/chapter17/intermediate/intermediate_design_patterns.dart

class User {
  final String name;
  final String email;
  final int age;
  const User({required this.name, required this.email, required this.age});
  @override
  String toString() => 'User($name, $email, $age)';
}

class UserBuilder {
  String _name = '';
  String _email = '';
  int _age = 0;

  UserBuilder name(String n) {
    _name = n;
    return this;
  }

  UserBuilder email(String e) {
    _email = e;
    return this;
  }

  UserBuilder age(int a) {
    _age = a;
    return this;
  }

  User build() => User(name: _name, email: _email, age: _age);
}

void runIntermediateDesignPatternExamples() {
  final user = UserBuilder()
      .name('Alice')
      .email('alice@mail.com')
      .age(28)
      .build();
  print('Built user: $user');
}

void main() {
  runIntermediateDesignPatternExamples();
}

File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter17/intermediate/IntermediateDesignPatterns.kt
package com.devlokos.masteringkotlin.chapters.chapter17.intermediate

/**
 * Capítulo 17 — Patrones de Diseño.
 * Nivel INTERMEDIO: Builder — construir User paso a paso con API fluida.
 */
fun main() {
    runIntermediateDesignPatternExamples()
}

data class User(val name: String, val email: String, val age: Int)

class UserBuilder {
    private var name: String = ""
    private var email: String = ""
    private var age: Int = 0

    fun name(n: String) = apply { name = n }
    fun email(e: String) = apply { email = e }
    fun age(a: Int) = apply { age = a }
    fun build(): User = User(name, email, age)
}

fun runIntermediateDesignPatternExamples() {
    val user = UserBuilder()
        .name("Alice")
        .email("alice@mail.com")
        .age(28)
        .build()
    println("Built user: $user")
}
